### PR TITLE
Support to train subsets of coco dataset

### PR DESCRIPTION
**Pull request by ZwwWayne**
There are some needs to train models on the subset of the COCO dataset, which only contain several categories, e.g., people, bicycles and cars. These models are taken as pre-trained models and further finetuned on small datasets such as KITTI. 

However, some bugs are reported when the users train models by themselves on subset of COCO dataset (e.g., in PR https://github.com/open-mmlab/mmdetection/issues/2288, https://github.com/open-mmlab/mmdetection/issues/2166, and https://github.com/open-mmlab/mmdetection/issues/424)
Therefore, this PR supports to set class names to be used when training on the subset of the COCO dataset, which could release the burden to subtract subsets before training. If the `class_names` is None, the data loader will use all the categories specified by the `self.CLASSES` by default, which uses all the data. If the `class_names` is a non-empty list, the dataset will use the `class_names` to find those images contains the specified categories and use them to train the model.

**Notice**: Currently, this behavior is only supported on the COCO dataset.

### CODE REVIEW

File: mmdet/datasets/coco.py [hunk 1]
```diff
@@ -32,13 +32,25 @@ class CocoDataset(CustomDataset):
                'vase', 'scissors', 'teddy_bear', 'hair_drier', 'toothbrush')
 
     def load_annotations(self, ann_file):
+        if not self.class_names:
```
Comment (xvjiarui) [inline mmdet/datasets/coco.py:35]: Why strip `_`?

File: mmdet/datasets/coco.py [hunk 2]
```diff
@@ -32,13 +32,25 @@ class CocoDataset(CustomDataset):
...
+            self.class_names = [res.replace('_', ' ') for res in self.CLASSES]
```
Comment (xvjiarui) [inline mmdet/datasets/coco.py:36]: Add comment about this.

**Comment by igonro** [comment]
Hi @ZwwWayne,

we have created a PR (#2408) that partially overlaps yours. We believe that the changes we want to add can also solve the use case you are dealing with in this Pull Request and it may be a good idea to generalize this solution to include also other use cases.

In order to fully match your use case, you would need to apply the following changes (the modified lines are marked with an asterisk, *):

In **datasets/coco.py:**

In order to filter the desired `self.cat_ids` (just like you do), we would have to modify the following in `load_annotations()` function:
```python
def load_annotations(self, ann_file):
    self.coco = COCO(ann_file)
*   self.cat_ids = self.coco.getCatIds(catNms=self.CLASSES)
    self.cat2label = {
        cat_id: i + 1
        for i, cat_id in enumerate(self.cat_ids)
    }
    self.img_ids = self.coco.getImgIds()
    img_infos = []
    for i in self.img_ids:
        info = self.coco.loadImgs([i])[0]
        info['filename'] = info['file_name']
        img_infos.append(info)
    return img_infos
```
Then, to filter the images that do not include labels of the desired categories, we could do the following in the `_filter_imgs()` function:
```python
def _filter_imgs(self, min_size=32):
    """Filter images too small or without ground truths."""
    valid_inds = []
    ids_with_ann = set(_['image_id'] for _ in self.coco.anns.values())
*   ids_with_valid_ann = set(_['image_id']
*                               for _ in self.coco.anns.values()
*                               if _['category_id'] in self.cat_ids)
    for i, img_info in enumerate(self.img_infos):
        if (self.filter_empty_gt and self.img_ids[i] not in ids_with_ann
*           ) or (self.custom_classes
*                   and self.img_ids[i] not in ids_with_valid_ann):
            continue
        if min(img_info['width'], img_info['height']) >= min_size:
            valid_inds.append(i)
    return valid_inds
```

In **datasets/custom.py:**

Also, to support the changes introduced in `filter_imgs()`, it would be necessary to modify the `__init__()` of the `CustomDataset` class as follows:
```python
def __init__(..., classes=None):
    [...]
    self.custom_classes = False

    if classes is not None:
        self.CLASSES = classes
        self.custom_classes = True
    [...]
```

I don't know if you think that this would cover your use case or there is something missing, but it would be nice if we can find an intermediate solution. Thank you!

File: configs/faster_rcnn_r50_fpn_1x_coco-subset.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
+        out_channels=256,
+        featmap_strides=[4, 8, 16, 32]),
+    bbox_head=dict(
+        type='SharedFCBBoxHead',
+        num_fcs=2,
+        in_channels=256,
+        fc_out_channels=1024,
+        roi_feat_size=7,
+        num_classes=81,
```
Comment (Dragonsson) [inline configs/faster_rcnn_r50_fpn_1x_coco-subset.py:41]: The num of classes you used still is 81.

**Comment by ZwwWayne** [comment]
> Hi @ZwwWayne,
> 
> we have created a PR (#2408) that partially overlaps yours. We believe that the changes we want to add can also solve the use case you are dealing with in this Pull Request and it may be a good idea to generalize this solution to include also other use cases.
> 
> In order to fully match your use case, you would need to apply the following changes (the modified lines are marked with an asterisk, *):
> 
> In **datasets/coco.py:**
> 
> In order to filter the desired `self.cat_ids` (just like you do), we would have to modify the following in `load_annotations()` function:
> 
> ```python
> def load_annotations(self, ann_file):
>     self.coco = COCO(ann_file)
> *   self.cat_ids = self.coco.getCatIds(catNms=self.CLASSES)
>     self.cat2label = {
>         cat_id: i + 1
>         for i, cat_id in enumerate(self.cat_ids)
>     }
>     self.img_ids = self.coco.getImgIds()
>     img_infos = []
>     for i in self.img_ids:
>         info = self.coco.loadImgs([i])[0]
>         info['filename'] = info['file_name']
>         img_infos.append(info)
>     return img_infos
> ```
> 
> Then, to filter the images that do not include labels of the desired categories, we could do the following in the `_filter_imgs()` function:
> 
> ```python
> def _filter_imgs(self, min_size=32):
>     """Filter images too small or without ground truths."""
>     valid_inds = []
>     ids_with_ann = set(_['image_id'] for _ in self.coco.anns.values())
> *   ids_with_valid_ann = set(_['image_id']
> *                               for _ in self.coco.anns.values()
> *                               if _['category_id'] in self.cat_ids)
>     for i, img_info in enumerate(self.img_infos):
>         if (self.filter_empty_gt and self.img_ids[i] not in ids_with_ann
> *           ) or (self.custom_classes
> *                   and self.img_ids[i] not in ids_with_valid_ann):
>             continue
>         if min(img_info['width'], img_info['height']) >= min_size:
>             valid_inds.append(i)
>     return valid_inds
> ```
> 
> To filter the annotations, you have to add the following code (exactly like you do) in `_parse_ann_info()` function (inside the for loop):
> 
> ```python
> if ann['category_id'] not in self.cat_ids:
>     continue
> ```
> 
> In **datasets/custom.py:**
> 
> Also, to support the changes introduced in `filter_imgs()`, it would be necessary to modify the `__init__()` of the `CustomDataset` class as follows:
> 
> ```python
> def __init__(..., classes=None):
>     [...]
>     self.custom_classes = False
> 
>     if classes is not None:
>         self.CLASSES = classes
>         self.custom_classes = True
>     [...]
> ```
> 
> I don't know if you think that this would cover your use case or there is something missing, but it would be nice if we can find an intermediate solution. Thank you!
> 
> **EDIT**: I forgot to include the code to filter the annotations.
> **EDIT2**: Hidden comment because after testing it, we realized that with minor fixes it also solves our use case.

Thanks for your kind suggestion. We plan to merge this PR after merging https://github.com/open-mmlab/mmdetection/pull/2408

File: mmdet/datasets/coco.py [hunk 4]
```diff
@@ -60,6 +69,26 @@ def _filter_imgs(self, min_size=32):
                 valid_inds.append(i)
         return valid_inds
 
+    def get_imgs_by_cat(self, imgIds=[], catIds=[]):
```
Comment (hellock) [inline mmdet/datasets/coco.py:72]: - `get_subset_by_classes()` looks better.
- `img_ids`, `class_ids`
- What is `img_ids` for? It seems to be unused.
- Use tuple or None as default values.
- Use the standard docsting style.

File: mmdet/datasets/custom.py [hunk 5]
```diff
@@ -156,6 +154,27 @@ def prepare_test_img(self, idx):
         self.pre_pipeline(results)
         return self.pipeline(results)
 
+    def get_classes(self, classes):
```
Comment (hellock) [inline mmdet/datasets/custom.py:157]: Make it a class method and return a list of class names.

File: mmdet/datasets/custom.py [hunk 6]
```diff
@@ -156,6 +154,27 @@ def prepare_test_img(self, idx):
...
+        """Get classes of current dataset
+
+        Args:
+            classes (None | str | tuple | list): If classes is None, use
```
Comment (hellock) [inline mmdet/datasets/custom.py:161]: `Sequence[str] | str | None`

File: mmdet/datasets/coco.py [hunk 7]
```diff
@@ -17,25 +17,34 @@
...
+        self.cat_ids = self.coco.getCatIds(catNms=self.CLASSES)
+
         self.cat2label = {cat_id: i for i, cat_id in enumerate(self.cat_ids)}
-        self.img_ids = self.coco.getImgIds()
+        # send cat ids to the get img id
+        # in case we only need to train on several classes
+        if self.custom_classes:
+            self.img_ids = self.get_imgs_by_cat(catIds=self.cat_ids)
+        else:
+            self.img_ids = self.coco.getImgIds()
```
Comment (hellock) [inline mmdet/datasets/coco.py:46]: Is the condition necessary?

File: mmdet/datasets/coco.py [hunk 8]
```diff
@@ -60,6 +66,26 @@ def _filter_imgs(self, min_size=32):
                 valid_inds.append(i)
         return valid_inds
 
+    def get_subset_by_classes(self, class_ids=None):
+        '''Get img ids that contain any category in class_ids.
+
+        Different from the coco.getImgIds, this function returns the id if
+        the img contains one of the cat rather than all.
```
Comment (hellock) [inline mmdet/datasets/coco.py:73]: Different from the `coco.getImgIds()`, this function returns the id if the img contains one of the categories rather than all.

File: configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco-ped-bicycle-car.py [hunk 9]
```diff
@@ -0,0 +1,7 @@
+_base_ = './faster_rcnn_r50_fpn_1x_coco.py'
+classes = ['person', 'bicycle', 'car']
```
Comment (hellock) [inline configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco-ped-bicycle-car.py:2]: Use tuple instead of list.

File: configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco-person.py [hunk 10]
```diff
@@ -0,0 +1,7 @@
+_base_ = './faster_rcnn_r50_fpn_1x_coco.py'
+classes = ('person')
+data = dict(
+    train=dict(classes=classes),
+    val=dict(classes=classes),
+    test=dict(classes=classes))
+load_from = 'https://s3.ap-northeast-2.amazonaws.com/open-mmlab/mmdetection/models/faster_rcnn_r50_fpn_1x_20181010-3d1b3351.pth'  # noqa
```
Comment (hellock) [inline configs/faster_rcnn/faster_rcnn_r50_fpn_1x_coco-person.py:7]: Remove `load_from`.

File: tests/test_dataset.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        classes=None,
+        test_mode=True,
+        img_prefix='VOC2007' if dataset == 'VOCDataset' else '')
+
+    assert custom_dataset.CLASSES == original_classes
+
+    # Test sending file path
+    import tempfile
+    import os.path as osp
+    tmp_file = tempfile.TemporaryDirectory()
```
Comment (hellock) [inline tests/test_dataset.py:51]: Use a temp file instead of a temp dir.

File: mmdet/datasets/custom.py [hunk 12]
```diff
@@ -156,6 +154,31 @@ def prepare_test_img(self, idx):
         self.pre_pipeline(results)
         return self.pipeline(results)
 
+    @classmethod
+    def get_classes(cls, classes):
```
Comment (hellock) [inline mmdet/datasets/custom.py:158]: Provide a default value `classes=None`.

File: tests/test_wrappers.py [hunk 13]
```diff
@@ -11,6 +11,11 @@
 
 
 def test_conv_2d():
```
Comment (hellock) [inline tests/test_wrappers.py:13]: test_conv2d

File: tests/test_wrappers.py [hunk 14]
```diff
@@ -26,7 +31,7 @@ def test_conv_2d():
         wrapper_out = wrapper(x_empty)
 
         # torch op with 3-dim input as shape reference
-        x_normal = torch.randn(3, in_cha, in_h, in_w)
+        x_normal = torch.randn(3, in_cha, in_h, in_w).requires_grad_(True)
```
Comment (hellock) [inline tests/test_wrappers.py:34]: `torch.randn(3, in_cha, in_h, in_w, requires_grad=True)`

File: mmdet/datasets/xml_style.py [hunk 15]
```diff
@@ -30,6 +30,26 @@ def load_annotations(self, ann_file):
             height = int(size.find('height').text)
             data_infos.append(
                 dict(id=img_id, filename=filename, width=width, height=height))
+
+        data_infos = self.get_subset_by_classes(data_infos)
```
Comment (hellock) [inline mmdet/datasets/xml_style.py:34]: Parsing the annotation files twice may be slow.

File: mmdet/datasets/custom.py [hunk 16]
```diff
@@ -156,6 +154,31 @@ def prepare_test_img(self, idx):
...
+
+        """
+        if classes is None:
+            return cls.CLASSES
+
+        if isinstance(classes, str):
+            # take it as a file path
+            class_names = mmcv.list_from_file(classes)
+        elif isinstance(classes, (tuple, list)):
+            class_names = classes
```
Comment (xvjiarui) [inline mmdet/datasets/custom.py:176]: We may raise an exception there.

File: mmdet/datasets/xml_style.py [hunk 17]
```diff
@@ -44,6 +75,8 @@ def get_ann_info(self, idx):
         labels_ignore = []
         for obj in root.findall('object'):
             name = obj.find('name').text
+            if name not in self.CLASSES:
+                continue
```
[the remark this answers is not in the document]
Comment (ZwwWayne) [reply mmdet/datasets/xml_style.py:67]: Yes. Because sometimes the annotation of an image contains instances of not only customized classes but also other classes. Those unused classes should be marked as background.

File: mmdet/datasets/custom.py [hunk 18]
```diff
@@ -156,6 +158,38 @@ def prepare_test_img(self, idx):
...
+        else:
+            raise ValueError('Unsupported type {} of classes.'.format(
+                type(classes)))
+
+        assert set(class_names).issubset(set(cls.CLASSES))
+        return class_names
+
+    def get_subset_by_classes(self):
+        pass
```
Comment (hellock) [inline mmdet/datasets/custom.py:191]: return self.data_infos

File: mmdet/datasets/custom.py [hunk 19]
```diff
@@ -156,6 +158,38 @@ def prepare_test_img(self, idx):
...
+        Args:
+            classes (Sequence[str] | str | None): If classes is None, use
+                default CLASSES defined by builtin dataset. If classes is a
+                string, take it as a file name. The file contains the name of
+                classes where each line contains one class name. If classes is
+                a tuple or list, override the CLASSES defined by the dataset.
+
+        """
+        if classes is None:
+            cls.custom_classes = True
```
Comment (hellock) [inline mmdet/datasets/custom.py:174]: True/False is wrong?

File: mmdet/datasets/xml_style.py [hunk 20]
```diff
@@ -30,8 +30,27 @@ def load_annotations(self, ann_file):
             height = int(size.find('height').text)
             data_infos.append(
                 dict(id=img_id, filename=filename, width=width, height=height))
+
         return data_infos
 
+    def get_subset_by_classes(self):
+        """Filter imgs by user-defined categories
+        """
+        clean_data_infos = []
```
Comment (hellock) [inline mmdet/datasets/xml_style.py:39]: subset_data_infos
